### ISSUE TITLE
CAPI controllers should keep ownerRefs on the current apiVersion

### ISSUE THREAD

**Issue by sbueringer** [kubernetes-sigs/cluster-api#7224]
I took a closer look at our ownerRefs and how they evolve (or not) after ClusterAPI upgrades. In some cases we bump the apiVersions automatically in some cases we don’t. Our cluster deletion reconciliation works independent of the apiVersion, but the Kubernetes garbage collection does not.

An Example:
* MD > MS > M is created with v1alpha3
* ownerRefs in MS and M are set with v1alpha3
* CAPI is upgraded a few times => ownerRefs stay on v1alpha3
* Eventually CAPI is upgraded to a version which doesn’t have v1alpha3
* => Kubernetes garbage collection is broken

Now let's take a look at how MachineDeployment deletion usually works:
* MD is deleted be the user
* kube-controller-manager garbage collector triggers deletion of MS when MD is gone
* MS is deleted
* kube-controller-manager garbage collector triggers deletion of Machines when MS is gone

Let's now assume the case from above. MS has an ownerRef from MD with an apiVersion that doesn't exist anymore (e.g. v1alpha3 in the example). The MD deletion now won't work anymore:
* MD is deleted
* MS still exists
  * The kube-controller-manager garbage collector is not able to “get” the MD in the old apiVersion

This seems like a bug to me and a blocker for removal of v1alpha3 apiVersions. I don't know for which resources exactly we depend on the Kubernetes garbage collection but I think we should keep all ownerRefs we set on the current apiVersion.

Proposed tasks:
* [ ] Probably makes sense to investigate if there is guidance on how to manage ownerRefs upstream
* [ ] audit all controllers/resources to figure out which ones are setting ownerRefs
* [ ] fix controllers to keep ownerRefs up-to-date (one example is the MD controller)
* [ ] broadcast this issue in office hours as it might be helpful for providers

It would be nice to have this implemented for v1.3.0 so we can consider dropping v1alpha3/v1alpha3 in the next releases.

/kind feature

**Comment by sbueringer** [comment]
cc @killianmuldoon Just in case you're interested

**Comment by killianmuldoon** [comment]
/assign

I'll take a look at this

**Comment by killianmuldoon** [comment]
It seems there is no upstream solution to this problem - it's been tracked in this issue for the last couple of years:
https://github.com/kubernetes/kubernetes/issues/96650

It would be great if we could get a solution that would also work in Controller Runtime, or in k/k itself to solve this problem across the community, but the first step here might just be to solve it in CAPI e.g. by upgrading references on every reconcile if required.

**Comment by sbueringer** [comment]
Yup - as discussed, I agree I think we have to start with a solution for CAPI.

**Comment by srm09** [comment]
Replicating this in CAPV as an investigation issue. 💯

**Comment by sbueringer** [comment]
/milestone v1.4

Given we are planning to implement this for v1.4

**Comment by killianmuldoon** [comment]
/reopen

This is no longer tested in CAPI e2e tests on main as we are no longer upgrading apiVersions in the clusterctl_upgrade tests. 

We should consider whether to try to continuously test this - e.g. by create a v1beta2 apiVersion for testing, or whether it's alright to leave this until there is a future apiVersion to upgrade to.

**Comment by sbueringer** [comment]
I think I wouldn't introduce an artificial apiVersion just to test this. Seems like a lot of effort

**Comment by killianmuldoon** [comment]
>Hm or wait, stupid idea. Could we just do the same as for the re-reconcile case, but instead of deleting the ownerRefs we just set them to some non-existent v1alpha1 version?

Not stupid at all - I think it should work perfectly.
